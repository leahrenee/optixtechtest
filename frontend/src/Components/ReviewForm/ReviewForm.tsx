import {
  Button,
  Container,
  TextField,
  Typography,
  styled,
} from "@mui/material";
import { useFormik } from "formik";
import * as yup from "yup";
import { usePostReview } from "../../hooks/usePostReview/usePostReview";
import { useState } from "react";
import { Movie } from "../../types";

interface Props {
  selectedMovie: Movie;
  setSelectedMovie: React.Dispatch<React.SetStateAction<Movie | undefined>>;
  handleFormSubmission: React.Dispatch<React.SetStateAction<string>>;
}

const classes = {
  form: "review-form",
};

const StyledContainer = styled(Container)(({ theme }) => ({
  padding: 20,

  [`& .${classes.form}`]: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    width: "100%",
    gap: 24,

    [theme.breakpoints.down("sm")]: {
      flexDirection: "column",
    },
    [theme.breakpoints.up("sm")]: {
      flexDirection: "row",
      justifyContent: "space-between",
    },
  },
}));

const ReviewForm = ({
  selectedMovie,
  setSelectedMovie,
  handleFormSubmission,
}: Props) => {
  const [reviewResponse, setReviewResponse] = useState<string>();

  // Post Hook
  const {
    isLoading: isSubmitReviewLoading,
    error: submitReviewError,
    postReview,
  } = usePostReview();

  const validationSchema = yup.object({
    review: yup
      .string()
      .required("Review is required to submit")
      .max(100, "Review must be less than 100 characters"),
  });

  const formik = useFormik({
    initialValues: {
      review: "",
    },
    validationSchema: validationSchema,
    onSubmit: async (value) => {
      const response = await postReview(value.review);

      if ((!isSubmitReviewLoading || !submitReviewError) && response) {
        setReviewResponse(response.message);
        handleFormSubmission(response.message);
      }

      setTimeout(() => {
        setSelectedMovie(undefined);
        setReviewResponse(undefined);
        handleFormSubmission("");
      }, 2000);
    },
  });

  return (
    <StyledContainer>
      <Typography style={{ marginBottom: "4px" }}>
        Please Enter Review for{" "}
        <span style={{ fontWeight: "bold", fontStyle: "italic" }}>
          {selectedMovie.title}
        </span>
      </Typography>
      <form
        onSubmit={formik.handleSubmit}
        className={classes.form}
        data-testid="review-form"
      >
        <TextField
          fullWidth
          variant="standard"
          id="review"
          name="review"
          label="Movie Review"
          value={formik.values.review}
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          error={formik.touched.review && Boolean(formik.errors.review)}
          helperText={formik.touched.review && formik.errors.review}
          disabled={!!reviewResponse}
        />
        <Button
          color="primary"
          variant="contained"
          type="submit"
          data-testid="submit-button"
          disabled={!!reviewResponse || !formik.isValid}
        >
          Submit
        </Button>
      </form>
    </StyledContainer>
  );
};

export default ReviewForm;
