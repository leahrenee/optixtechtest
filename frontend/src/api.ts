const baseUrl = "http://localhost:3000";

export const getData = async <T>(path: string): Promise<T> => {
  let data;

  try {
    const res = await fetch(`${baseUrl}/${path}`);

    if (res.ok) {
      data = await res.json();
    } else {
      if (res.status === 404) throw new Error("404, Not found");
      if (res.status === 500) throw new Error("500, internal server error");
      throw new Error(`Server responded with ${res.status}`);
    }
  } catch (error) {
    throw new Error(`Fetching failed: ${error}`);
  }

  return data;
};

export const postData = async <T, R>(path: string, data: T): Promise<R> => {
  let reviewResponse;

  try {
    const res = await fetch(`${baseUrl}/${path}`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        body: JSON.stringify(data),
      },
    });

    if (res.ok) {
      reviewResponse = await res.json();
    } else {
      if (res.status === 404) throw new Error("404, Not found");
      if (res.status === 500) throw new Error("500, internal server error");
      throw new Error(`Server responded with ${res.status}`);
    }
  } catch (error) {
    console.error("Fetch", error);
  }

  return reviewResponse;
};
