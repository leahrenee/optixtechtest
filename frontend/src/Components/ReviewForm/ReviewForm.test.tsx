import { describe, test } from "@jest/globals";
import { fireEvent, render, waitFor } from "@testing-library/react";
import ReviewForm from "./ReviewForm";
import {
  mockMoviesData,
  mockPostReviewRequest,
  mockPostReviewResponse,
} from "../../testData";
import { postData } from "../../api";

jest.mock("../../api");
const mockPostData = postData as jest.Mock;

const component = (
  <ReviewForm
    selectedMovie={mockMoviesData[0]}
    setSelectedMovie={jest.fn()}
    handleFormSubmission={jest.fn()}
  />
);

describe("ReviewForm", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("Should match default snapshot", async () => {
    const { asFragment } = render(component);

    expect(asFragment()).toMatchSnapshot();
  });

  test("Should submit form with correct review", async () => {
    mockPostData.mockReturnValueOnce(Promise.resolve(mockPostReviewResponse));

    const { getByTestId, getByLabelText } = render(component);

    const reviewInput = getByLabelText("Movie Review");
    const submitButton = getByTestId("submit-button");

    fireEvent.change(reviewInput, {
      target: { value: mockPostReviewRequest.review },
    });

    fireEvent.click(submitButton);

    await waitFor(() =>
      expect(mockPostData).toHaveBeenCalledWith("submitReview", {
        review: mockPostReviewRequest.review,
      })
    );
  });
});
