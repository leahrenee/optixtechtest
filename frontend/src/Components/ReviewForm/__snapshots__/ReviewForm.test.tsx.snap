// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewForm Should match default snapshot 1`] = `[Function]`;
